lib/Sema/SemaBounds.cpp

```diff
@@ -578,6 +578,237 @@ namespace {
+      if (!CheckValuePreservingCast(E, ToType)) {
+        // it's non-value-preserving, stop
+        return;
+      }
+
+      Expr *Needle = E->getSubExpr();
+      QualType NeedleTy = Needle->getType();
+      while (true) {
+        Needle = Needle->IgnoreParens();
+        NeedleTy = Needle->getType();
+
+        if (Needle->isNullPointerConstant(S.Context, Expr::NPC_NeverValueDependent))
+          // 3a. We've got to a null pointer, so this cast is allowed, stop
+          return;
+
+        if (NeedleTy->isCheckedPointerPtrType()) {
+          // 1. We've found something with ptr<> type, check compatibility.
+
+          bool types_are_compatible = S.Context.typesAreCompatible(ToType, NeedleTy,
+                                                                   /*CompareUnqualified=*/false,
+                                                                   /*IgnoreBounds=*/false);
+          if (!types_are_compatible) {
+            // 1b) it is incompatible with ToType, add an error
+            S.Diag(Needle->getExprLoc(), diag::err_cast_to_checked_fn_ptr_from_incompatible_type)
+              << ToType << NeedleTy << true
+              << E->getSourceRange();
+          }
+
+          // We can stop here, as we've got back to something of checked ptr<> type. 
+          // CheckDisallowedFunctionPtrCasts will be called on any sub-expressions if they
+          // are potentially problematic casts to checked ptr<> types. 
+          return;
+        }
+
+        // If we've found a cast expression...
+        if (CastExpr *NeedleCast = dyn_cast<CastExpr>(Needle)) {
+          // 2. check if the cast is value preserving
+          if (!CheckValuePreservingCast(NeedleCast, ToType)) {
+            // it's non-value-preserving, stop
+            return;
+          }
+
+          // it is value-preserving, continue descending
+          Needle = NeedleCast->getSubExpr();
+          NeedleTy = Needle->getType();
+          continue;
+        }
+
+        // If we've found a unary operator (such as * or &)...
+        if (UnaryOperator *NeedleOp = dyn_cast<UnaryOperator>(Needle)) {
+          // 2. Check if the operator is value-preserving.
+          //    Only addr-of (&) and deref (*) are with function pointers
+          if (!CheckValuePreservingCastLikeOp(NeedleOp, ToType)) {
+            // it's not value-preserving, stop
+            return;
+          }
+
+          // it is value-preserving, continue descending
+          Needle = NeedleOp->getSubExpr();
+          NeedleTy = Needle->getType();
+          continue;
+        }
+
+        // If we've not found a cast or a cast-like operator, 
+        // then we stop descending
+        break;
+      }
+
+      // This feels like a terrible hack, but it's the only way that the type compatibility check
+      // immediately below will pass
+      if (NeedleTy->isFunctionType()) {
+        NeedleTy = S.Context.getPointerType(NeedleTy, CheckedPointerKind::Ptr);
```

Comment:
This needs further discussion. It is part of the check that the `ToType` matches the thing we've got to.